Add the Stokes Forces Implementation for the Surface Bands wave coupling method

alperaltuntas opened the pull request:
This PR adds the Stokes Forces implementation by @breichl and some additional refactoring and bug fixes. 

The summary of Stokes Forces changes (@breichl):

 - Introduction of the stokes forces based on the 3d Stokes drift profile. The 3d Stokes drift profile is obtained via the partitioned Stokes drift components exported from WW3. The added forces are the duS/dt term, the updated pressure gradient force, and the the vorticity modifications. For the duS/dt term, the increment tendency is computed whenever the Stokes drift is updated. This term is applied on the dynamics time step. The pressure gradient force is being calculated outside of the model's pressure solver, and  the Stokes-shear terms at cell centers and then integrating and taking gradients to estimate a pressure gradient tendency anomaly. This isn't the correct way of incorporating the pressure gradient force, so will be corrected soon. 
 
The summary of refactorings and bugfixes (@alperaltuntas) :

- Add the ability to import fields with ungridded dimensions in NUOPC cap.
- Add a more flexible and concise way of exporting partitioned stokes drift components from WW3. This is based on utilizing the NUOPC ungridded dims feature. The new way of exporting these fields is encapsulated with `if (cesm_coupled)` checks. However, when the anticipated unification of EMC and NCAR WW3 nuopc caps is accomplished, the old way of exporting the partitioned stokes drift components may be removed to get rid of duplicate code.
- Fixed the exact restart test failures by adding the 3d stokes drift field to the restart file. To do so,  minor refactoring is done in wave_interface module and calling driver modules.
- other bugfixes: remove unnecessary ampersands from variable declarations. enclose stokes vorticity diag computations with Stokes_VF conditional. Do not call get_Langmuir_Number if lamult is already provided via ww3. Fix openmp test by adding Stokes_VT to shr clause.

Commit:
Adding Stokes forces implementation and diags.

- Add Craik-Leibovich terms, vorticity and dynamic pressure components.
- Add exploratory Stokes ddt version in MOM_wave_interface and MOM.F90
- Add tendency diagnostics for Stokes terms w/ control and option for diagnostic only Stokes force computation in MOM_CoriolisAdv.

Commit:
Merge branch 'dev/gfdl' into Stokes_Forces

Commit:
Finish incomplete merge in MOM_wave_interface.

Commit:
Remote application of Stokes tendency on thermodynamic step

- This only makes sense to add on the dynamics timestep.

Commit:
Updating name of Stokes PGF routine to be more descriptive.

Commit:
Merge branch 'dev/gfdl' into Stokes_Forces

alperaltuntas committed:
Merge remote-tracking branch 'breichl/Stokes_Forces' into pstokes_wave_cpl_mehtod

alperaltuntas committed:
Merge branch 'merge_ww3_coupling_vr12ma' into pstokes_wave_cpl_mehtod

alperaltuntas committed:
In NUOPC cap, add ability to import fields with ungridded dimensions

alperaltuntas committed:
remove unnecessary ampersands from variable declarations

alperaltuntas committed:
enclose stokes vorticity diag computations with Stokes_VF conditional

alperaltuntas committed:
enclose stokes vorticity diag computations with Stokes_VF conditional

alperaltuntas committed:
Merge branch 'surfbands_refactor' of https://github.com/NCAR/MOM6 into surfbands_refactor

alperaltuntas committed:
Do not call get_Langmuir_Number if lamult is already provided via ww3

alperaltuntas committed:
fix doxygen errors in surfbands_refactor branch

alperaltuntas committed:
changes in the nuopc cap to support arbitrary stokes bands
With these changes, arbitrary number of of partitioned stokes drift components
may be imported from ww3 when coupled within cesm. After the planned unification
of ww3 nuopc caps of CESM and EMC, the old (fixed) approach may be removed.

alperaltuntas committed:
refactoring and changes to write stokes drift
profile to restart file when surfbands wave coupling mode is on.

alperaltuntas committed:
Merge branch 'dev/ncar' into surfbands_refactor

alperaltuntas committed:
For the drivers calling MOM_wave_interface_init,
add the optional waves_csp arg to initialize_MOM
calls so as to ensure that the waves_csp is allocated.
This commit also has several misc doxygen fixes.

alperaltuntas committed:
initialize wave option flags to false

alperaltuntas committed:
Fix openmp test by adding Stokes_VT to shr clause
in MOM_CoriolisAdv. Also remove the Passive_Stokes_VF
block since it's never set to true.

alperaltuntas commented on src/core/MOM_CoriolisAdv.F90:
```diff
@@ -284,15 +293,41 @@ subroutine CorAdCalc(u, v, h, uh, vh, CAu, CAv, OBC, AD, G, GV, US, CS)
+        dudy(I,J) = (u(I,j+1,k)*G%dxCu(I,j+1) - u(I,j,k)*G%dxCu(I,j))
+      enddo; enddo
+    endif
     do J=Jsq-1,Jeq+1 ; do I=Isq-1,Ieq+1
```
Just noting here that the following do loop (lines 331-334) needs to be removed, i think.

alperaltuntas committed:
merge with latest dev/ncar and resolve conflicts

alperaltuntas committed:
remove duplicate assignments of dvdx and dudy. These duplicate assignments incorrectly override dvdx and dudy when surfbands wave coupling is on.

breichl commented on src/user/MOM_wave_interface.F90:
```diff
@@ -1441,7 +1507,7 @@ subroutine CoriolisStokes(G, GV, dt, h, u, v, WAVES, US)
 
   do k = 1, GV%ke
-    do j = G%jsc, G%jec
+do j = G%jsc, G%jec
```
I think this was not intended.

breichl commented on src/user/MOM_wave_interface.F90:
```diff
@@ -1537,6 +1746,51 @@ subroutine Waves_end(CS)
+  if (trim(wave_method_str)== trim(SURFBANDS_STRING)) then
+    vd(1) = var_desc("US_x", "m s-1", "3d zonal Stokes drift profile")
+    vd(2) = var_desc("US_y", "m s-1", "3d meridional Stokes drift profile")
+    call register_restart_field(CS%US_x(:,:,:), vd(1), .true., restart_CSp)
```
The logical here has to be set to .false. to work in symmetric memory mode.  I do not understand why.

Commit:
Fixes for Stokes time tendency terms

- Adding logical to prevent using terms if not allocated
- Adding optional diagnostic only output of term
- Moving time increment to within dynamics step instead of in main step MOOM loop.
- Correcting time tendency term to reflect time between Stokes drift updates and extent of dynamics time loop.

alperaltuntas committed:
fixes in MOM_wave_interface array indices

alperaltuntas committed:
More changes for surfbands wave coupling:
- for ustkb and vstkb, call pass_var, instead of pass_vector because
they are on h grd.
-remove unused ustk0 and vstk0 arrays.
- correct the order of wave-related get param & allocate_forcing calls
in nuopc cap.
- add lamult flag to allocate_forcing_by_group because lamult should
only be allocated if wave_method=="EFACTOR".

Commit:
Updates	to Stokes drift	terms, primarily the pressure gradient routine.

- Replaces Stokes-induced pressure anomaly gradient routine with more accurate method that explicitly integrates the Stokes-shear force contribution to the pressure.  Includes diagnostics for the pressure anomaly to verify.
- Updates the Stokes time derivative to only be updated on dynamics time-steps.  Adds additional storage of previous step that is also only updated on the dynamics time-steps.
- Update so that the surface Stokes drift output when using the exponential surfbands option is the surface Stokes drift and not averaged over a layer near the surface.
- Minor rename for clarity in Update_Surface_Waves using data_override to clarify the time that wave terms are computed.
- Adds diagnostic tracking Stokes time tendency term for verification.
- Updates Stokes diagnostic names for conformity.
- Fix allocation of CS%STKy0 from u grid to v grid.

Commit:
Fixing diagnostic mode for vortex force correction term.

- The model's Coriolis/acceleration term was incorrect when the Stokes vortex form correction was in diagnostic mode.  The correct calculations are restored so it can be run in diagnostic mode and reproduce with the model when the setting is turned off entirely.

Commit:
Fixing restart issue with Stokes time tendency term.

- The term us_x and us_y needed to be replaced with us_x_prev and us_y_prev to store the previous timestep Stokes drift in order to have reproducing code with restart files.

Commit:
Merge remote-tracking branch 'NCAR/surfbands_refactor' into BGR_Stokes_DDT_Fixes

alperaltuntas committed:
fix line lenght error in MOM_wave_interface

alperaltuntas commented:
@gustavo-marques this is ready to be tested and merged.

gustavo-marques commented:
@alperaltuntas, I have verified that this PR does not change answers for the MOM6-examples test suit w.r.t. the latest main (https://github.com/NCAR/MOM6/commit/9cb93044ef06a929171b5bd326a7c5c06aa72316). Were you able to run the CESM tests?

alperaltuntas commented:
@gustavo-marques Yes, the aux_mom test suite passes.
